karoo_gp/base_class.py

```diff
@@ -211,68 +191,138 @@ def error(self, msg, display={'i', 'g', 'm', 'db'}):
+                    self.tourn_size,
+                    self.evolve_repro,
+                    self.evolve_point,
+                    self.evolve_branch,
+                    self.evolve_cross,
                 )
-                self.log('\n Copy gp.population_b to gp.population_a\n')
 
-            if self.mode == 's':
-                # (s)erver mode - termination with completiont of prescribed run
+                # Evaluate new generation
+                self.population.evaluate(X_train, y_train, X_train_hash)
+
+                # Add best score to history
+                for k, v in self.score(self.X_test, self.y_test).items():
+                    self.history[k].append(v)
+
+            if self.mode == 's':  # (s)erver mode: terminate after run
                 menu = 0
-            else:
-                # (d)esktop mode - user is given an option to quit, review,
-                # and/or modify parameters; 'add' generations continues the run
-                self.log('\n\t\033[32m Enter \033[1m?\033[0;0m\033[32m to review '
-                         'your options or \033[1mq\033[0;0m\033[32muit\033[0;0m')
+            else:                 # (d)esktop mode: pause after run
+                self.log('Enter ? to review your options or q to quit')
                 menu = self.pause()
 
+    def predict(self, X):
+        """Return predicted y values for X using the fittest tree
+
+        * Pass the fittest tree to batch_predict as a list (of 1)
+        * Return the first result
+        """
+        fittest_tree = self.population.fittest()
+        return self.batch_predict(X, [fittest_tree])[0]
+
+    def batch_predict(self, X, trees, X_hash=None):
+        """Return predicted values for y given X for a list of trees
+
+        * If encoder (e.g. MultiClassifier), decode predictions
+        * If precision is specified, round predictions to precision
+        """
+        y = self.engine.predict(trees, X, X_hash)
+        if self.encoder is not None:
+            y = self.encoder.decode(y)
+        if self.precision is not None:
+            y = np.round(y, self.precision)
+        return y
+
+    def score(self, X, y):
+        """Return score of the fittest tree on X and y"""
+        return self.calculate_score(self.predict(X), y)
+
+    def calculate_score(self, y_pred, y_true):
+        """Return a dict with hthe results of each scoring function"""
+        return {label: fx(y_true, y_pred)
+                for label, fx in self.scoring.items()}
+
     def fx_karoo_terminate(self):
         '''
         Terminates the evolutionary run (if yet in progress),
         saves parameters and data to disk, and cleanly returns
         the user to karoo_gp.py and the command line.
-
-        Called by: fx_karoo_gp() and fx_karoo_pause_refer()
-
-        Arguments required: none
         '''
+        kernel = {
+            RegressorGP: 'r', MultiClassifierGP: 'c', MatchingGP: 'm'
+        }[type(self)]
```

Comment:
Don't you already have this in `self.tree_type`?
This will also break if people add new types.

Reply:
`tree_type` (full, grow, ramped) != `kernel` (regression, class.., matching)

Ya, this is a workaround to make the tests pass. We don't use the kernel now except in `karoo-gp.py`. It's used below to 
(1) determine what to output for score, which we know from the `.score` dict, and 
(2) save a record of the model used. Can the class use its own name for saving (e.g. MatchingGP) instead of the kernel?